src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+
+                if (currentItemPriority > bestItemPriority)
+                {
+                    bestItem = chosenItem;
+                }
+            }
+
+            return bestItem;
+        }
+
+        private EditorCompletion.FilteredCompletionModel HandleAllItemsFilteredOut
+            (EditorCompletion.CompletionTriggerReason triggerReason,
+            ImmutableArray<EditorCompletion.CompletionFilterWithState> filters,
+            ImmutableArray<EditorCompletion.CompletionFilter> activeFilters,
+            CompletionRules completionRules)
+        {
+            EditorCompletion.UpdateSelectionHint selection;
+            if (triggerReason == EditorCompletion.CompletionTriggerReason.Insertion)
+            {
+                // If the user was just typing, and the list went to empty *and* this is a 
+                // language that wants to dismiss on empty, then just return a null model
+                // to stop the completion session.
+                if (completionRules.DismissIfEmpty)
+                {
+                    return null;
+                }
+
+                selection = EditorCompletion.UpdateSelectionHint.NoChange;
+            }
+            else
+            {
+                // If the user has turned on some filtering states, and we filtered down to
+                // nothing, then we do want the UI to show that to them.  That way the user
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                selection = activeFilters.Length == 0 ? EditorCompletion.UpdateSelectionHint.SoftSelected : EditorCompletion.UpdateSelectionHint.NoChange;
+            }
+
+            return new EditorCompletion.FilteredCompletionModel(ImmutableArray<EditorCompletion.CompletionItemWithHighlight>.Empty, 0, filters, selection, centerSelection: true, uniqueItem: default);
+        }
+
+        private bool MatchesFilterText(
+            string itemFilterText,
+            string itemDisplayText,
+            int matchPriority,
+            string filterText,
+            EditorCompletion.CompletionTriggerReason triggerReason)
+        {
+            // For the deletion we bake in the core logic for how matching should work.
+            // This way deletion feels the same across all languages that opt into deletion
+            // as a completion trigger.
+
+            // Specifically, to avoid being too aggressive when matching an item during
+            // completion, we require that the current filter text be a prefix of the
+            // item in the list.
+
+            if (triggerReason == EditorCompletion.CompletionTriggerReason.Deletion || triggerReason== EditorCompletion.CompletionTriggerReason.Backspace)
```

Comment:
where did the logic for checking "filterReason" go.

Reply:
Editor provides us only the current trigger. So, in some intermediate versions we copied it twice: as a trigger reasons and as a filter reason. Then, we passed it through all the stack and made same comparisons here.

This looked bad. So, I removed the second reason at least for a while.

We may consider restoring it in the future if we will meet scenarios requiring this: we still have some unit tests not supported in the new completion world.

---
In reply to: [222089372](https://github.com/dotnet/roslyn/pull/29016#discussion_r222089372) [](ancestors = 222089372)